core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -119,6 +115,11 @@ void update(ShardRouting oldShard, ShardRouting newShard) {
         assert previousValue == oldShard : "expected shard " + previousValue + " but was " + oldShard;
     }
 
+    void remove(ShardRouting shard) {
+        ShardRouting previousValue = shards.remove(shard.shardId());
+        assert previousValue == shard : "expected shard " + previousValue + " but was " + shard;
```

Comment:
I prefer this to be a soft check as it only verifies if we're using the internal API properly.